### ISSUE TITLE
More distribution functions

### ISSUE THREAD

**Issue by rstub** [daqana/dqrng#22]
Many distribution functions are supported via the [stats package](https://stat.ethz.ch/R-manual/R-patched/library/stats/html/Distributions.html) and [add-on packages](https://cran.r-project.org/view=Distributions) together with Rs RNG. It would be good to support a greater selection of distributions together with the faster RNGs from dqrng.

Directly usable:

* https://www.boost.org/doc/libs/1_69_0/doc/html/boost_random/reference.html#boost_random.reference.distributions

**Comment by rstub** [comment]
Potentially usable (c.f. https://stackoverflow.com/a/54143984/8416610):

* https://cran.r-project.org/package=rgen
* https://cran.r-project.org/package=RcppDist

@coatless, @duckmayr: Would any of you be interested in supporting external RNGs?

Generalized Ziggurat: https://github.com/DiscreteLogarithm/Zest

**Comment by duckmayr** [comment]
@rstub Potentially. I can look into this around March 23.

**Comment by coatless** [comment]
@rstub more than content with adding new routines into `rgen`. That said, the `rgen` package was created prior to the development of @kthohr's [`stats`](https://github.com/kthohr/stats) as I needed _R_ sampling routines that were not exported.

**Comment by rstub** [comment]
@coatless I don't think that additional routines would be necessary. I think it should work to add additional arguments to the existing functions. So instead of calling a fixed `norm_rand` or `unif_rand`, a function is called that has been passed in as argument and defaults to the currently used function. This way the existing functions should be usable as before but offer the possibility to change the random number source.

Concerning the `stats` library. One problem is that it relies on (at least) `std::uniform_distribution` and `std::normal_distribution`. This makes it unsuitable for usage in an R package, since WRE clearly says that one should not use the `random` header from C++11. And the *implementation defined* distribution functions are a good reason for this.

**Comment by rstub** [comment]
Distributions that are used in packages making use of dqrng:
* log-normal
* gamma

**Comment by jessekps** [comment]
Thanks for considering the gamma. For completenss sake, I actually use a couple more, but they are rather trivial extensions of a potential rgamma. Nevertheless you might be able to improve on them.

```
double rchisq(dqrng::xoshiro256plus& lrng, const double df)
{
	return 2 * rgamma(lrng, df/2, 1.0);
}
	
double rinvchisq(dqrng::xoshiro256plus& lrng, const double df, const double scale)
{
	double z = rchisq(lrng, df);
	if(z==0) z= 1e-100;
	return (df*scale) / z;
}

double rbeta(dqrng::xoshiro256plus& lrng, const double alpha, const double beta)
{
	double x = rgamma(lrng, alpha, 1.0);
	double y = rgamma(lrng, beta, 1.0);
	
	return x/(x+y);
}
```

And also a dirichlet, but the implementation is a little specific to my package and not sure how you would want to generalize it since it would need a pointer, vector or iterator of some kind.
